tools/migration/internal/runner.go

```diff
@@ -26,27 +41,169 @@ type Migration interface {
 	Validate(oldRepoPath, newRepoPath string) error
 }
 
-// MigrationRunner represent a migration command
+// MigrationRunner represents a migration command
 type MigrationRunner struct {
-	logger     *Logger
-	command    string
+	// logger logs to stdout/err and a logfile.
+	logger *Logger
+
+	// command is the migration command to run, passed from the CLI
+	command string
+
+	// oldRepoOpt is the value of --old-repo passed  from the CLI
 	oldRepoOpt string
-}
 
-// NewMigrationRunner builds a MirgrationRunner for the given command and repo options
-func NewMigrationRunner(logger *Logger, command, oldRepoOpt string) *MigrationRunner {
-	// TODO: Issue #2585 Implement repo migration version detection and upgrade decisioning
+	// newRepoPath is where the to-be-migrated/migrated repo is located
+	newRepoPath string
 
+	// MigrationsProvider is a dependency for fetching available migrations
+	// to allow unit tests to supply test migrations without creating test fixtures.
+	MigrationsProvider func() []Migration
+}
+
+// NewMigrationRunner builds a MigrationRunner for the given command and repo options
+func NewMigrationRunner(logger *Logger, command, oldRepoOpt, newRepoOpt string) *MigrationRunner {
 	return &MigrationRunner{
-		logger:     logger,
-		command:    command,
-		oldRepoOpt: oldRepoOpt,
+		logger:             logger,
+		command:            command,
+		oldRepoOpt:         oldRepoOpt,
+		newRepoPath:        newRepoOpt,
+		MigrationsProvider: DefaultMigrationsProvider,
 	}
 }
 
 // Run executes the MigrationRunner
 func (m *MigrationRunner) Run() error {
-	// TODO: Issue #2595 Implement first repo migration
+	repoVersion, err := m.repoVersion(m.oldRepoOpt)
+	if err != nil {
+		return err
+	}
+	if repoVersion == m.getTargetMigrationVersion() {
+		return fmt.Errorf("binary version %d = repo version %d; migration not run", repoVersion, m.getTargetMigrationVersion())
+	}
+
+	var mig Migration
+	if mig, err = m.getValidMigration(repoVersion); err != nil {
+		return fmt.Errorf("migration check failed: %s", err.Error())
```

Comment:
If getValidMigration fails, it could be that the version file was corrupt or there is >1 valid migration, in which case there is a migration bug.  By the time of this check, we would already have detected whether the repo is up-to-date (which is not an error case), so we know migration is needed. If we can't find one and only one that applies, that sounds like a valid error to me.